README
NAME
    Env::Elope - Configure projects in multiple environments

VERSION
    version 0.001

SYNOPSIS
        # lib/My/Project.pm
        use Env::Elope qw( dist_env_dir );
        my $config = dist_env_dir( 'My-Project' )->child( 'db.yml' );

        # share/env/dev/db.yml
        db_server: dev.example.com

        # share/env/prd/db.yml
        db_server: example.com

        # On a command line:
        $ elope dev run_my_project
        $ elope prd run_my_project

DESCRIPTION
    Elope helps to manage multiple environments through configuration
    directories and environment variables. Using the "elope" command-line
    script, a program can be run in any environment, choosing all the right
    configuration for that environment.

  SETUP
    To set up Elope, you need:

    A config directory named for each environment
        The configuration directories should be in each distribution's
        "share/env" directory and should have the same name as the
        environment they configure. So "share/env/dev" will be our "dev"
        environment, and "share/env/prod" our "prod" environment.

    The "elope" command-line utility
        The elope utility will source in the envrc file for the current
        environment and then run the command. The elope utility sets the
        "ELOPE_ENV" environment variable that chooses which config directory
        to use.

    An envrc file named for each environment
        After installing the elope utility, you must make a directory for
        your envrc files. If you installed elope to /opt/bin/elope, you
        should make a directory /opt/etc/envrc and put your envrc files
        there.

        An envrc file should be named the same as the environment, and
        should export environment variables like "PATH" and
        "LD_LIBRARY_PATH", or maybe configure a Perl local::lib.

            # /opt/etc/envrc/dev
            # Now that 5.20.2 is released, lets use it in dev to try it
            PATH=/usr/perlbrew/perls/5.20.2/bin:$PATH
            eval $( perl -Mlocal::lib=/opt/envs/dev )

            # /opt/etc/envrc/prod
            # We haven't upgrade to 5.20.2 yet in prod
            PATH=/usr/perlbrew/perls/5.20.1/bin:$PATH
            eval $( perl -Mlocal::lib=/opt/envs/prod )

  USING
    Now that we have our envrc and our config dirs, we can run our scripts
    with the elope command:

        # Install one of our modules in dev
        $ elope dev perl Build.PL
        $ ./Build install

        # Run its tests using the dev configuration
        $ elope dev prove -lr t

        # Run our server
        $ elope dev morbo bin/my-mojo-app -l 'http://*:8080'

    If we switch "dev" to "prod" in any of the above commands, we are
    running them in production.

SUBROUTINES
  dist_env_dir DIST, OPTIONS
    Get the configuration directory for the given "DIST". Returns a
    Path::Tiny object with the given dist path.

    Configuration directories are located in a distribution's "share/env/"
    directory. Which directory is chosen is based on the "ELOPE_ENV"
    environment variable.

    "OPTIONS" is a set of key/value pairs for options with the following
    keys:

        env     - Set the environment name, overriding the current environment

AUTHOR
    Doug Bell <preaction@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2015 by Doug Bell.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Changelog for Env-Elope

{{ $NEXT }}

 - Initial release


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.023.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Configure projects in multiple environments",
  "dist_author" => [
    "Doug Bell <preaction\@cpan.org>"
  ],
  "dist_name" => "Env-Elope",
  "dist_version" => "0.001",
  "license" => "perl",
  "module_name" => "Env::Elope",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "File::Share" => 0,
    "Sub::Exporter" => 0,
    "perl" => "5.008"
  },
  "script_files" => [
    "bin/elope"
  ],
  "test_requires" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Deep" => 0,
    "Test::Differences" => 0,
    "Test::Exception" => 0,
    "Test::More" => "1.001005"
  }
);


my %fallback_build_requires = (
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.28",
  "Test::Deep" => 0,
  "Test::Differences" => 0,
  "Test::Exception" => 0,
  "Test::More" => "1.001005"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.023.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Configure projects in multiple environments",
  "AUTHOR" => "Doug Bell <preaction\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.28"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "DISTNAME" => "Env-Elope",
  "EXE_FILES" => [
    "bin/elope"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Env::Elope",
  "PREREQ_PM" => {
    "File::Share" => 0,
    "Sub::Exporter" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Deep" => 0,
    "Test::Differences" => 0,
    "Test::Exception" => 0,
    "Test::More" => "1.001005"
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Share" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.28",
  "Sub::Exporter" => 0,
  "Test::Deep" => 0,
  "Test::Differences" => 0,
  "Test::Exception" => 0,
  "Test::More" => "1.001005"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



